source/common/upstream/subset_lb.cc

```diff
@@ -215,6 +215,20 @@ void SubsetLoadBalancer::processSubsets(
 // new subsets as necessary.
 void SubsetLoadBalancer::update(uint32_t priority, const HostVector& hosts_added,
                                 const HostVector& hosts_removed) {
+  // It's possible that metadata changed, without hosts being added nor removed.
+  // If so, we need to go over all hosts to check for new or unused subsets.
+  if (!hosts_added.size() && !hosts_removed.size()) {
+    for (auto& host_set : original_priority_set_.hostSetsPerPriority()) {
+      // Prevent endless recursion.
+      if (!host_set->hosts().size()) {
+        continue;
+      }
+      update(host_set->priority(), host_set->hosts(), {});
```

Comment:
I don't believe this is correct, because you haven't removed the copy of the hosts that are stored in the hierarchy with the old metadata. So a given host will still be mapped to it's previous subset. Not clear how you fix that, except to write some code to manually remove all the subsets and rebuild them all, which seems like a pain.

I thought I read a comment that said having updateDynamicHostList return true caused a full rebuild of the load balancer? If that's true, it seems like this shouldn't matter, but I guess it actually does just attempt to update the hosts. Given that it actually comes through as empty update, this is an argument for having metadata changes look like host removal and addition.

Reply:
Hmm, what #3810 will trigger is a call to `HostSetImpl::updateHosts()`:

https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/eds.cc#L203

which eventually ends up calling `SubsetLoadBalancer::update()` via `runUpdateCallbacks()`. 

I think we'll refilter hosts in `SubsetLoadBalancer::HostSubsetImpl::update()`:

https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/subset_lb.cc#L454

which should give us an updated subset... no?

Reply:
@zuercher the rebuild only happens for dumb LBs; for subset LB, the above `update()` is invoked and it gets to do the more performant thing with `hosts_added/hosts_removed`. 

I do think this would be a lot simpler to model as a remove/add; @rgs1 what was the concern with this? Forcing a drain and stats reset? Could we just remove/add from the PoV of the cluster membership (i.e. in the `hosts_added/hosts_removed` parameters) while not actually nuking the underlying host object in https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/upstream_impl.cc#L682?

Reply:
the concern within #3810? a few things: draining, stats reset, making `updateDynamicHostList()` too complex... 

i can't come up with a super strong argument now, but i suspect we'll have more friction in the future if we conflate metadata updates with adds/removals.... 

i'll think about it a bit more, but i am also a bit tight on time budget, so happy to explore other alternatives.

Reply:
Can we implement this as a pure host add? It looks like the delta in your PR is just doing this; you're basically treating `hosts_added` as `hosts_added_or_modified` in the recursion. Should we rename the field to indicate this?

Reply:
Yeah, actually — I think we should move this into `addMemberUpdateCb()`:

https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/subset_lb.cc#L41

which will make things more obvious. I'll push something in a bit.